### ISSUE TITLE
MIPS64r6 RELEASE doesn't compile with CLANG

### ISSUE THREAD

**Issue by ibogosavljevic** [v8mips/v8mips#193]
Starting from https://v8.mips.com/bbv8/builders/Target%20-%20MIPS64%20-%20edge%20-%20clang%2Cmipsr6%20-%20builder/builds/626, MIPS64r6 RELEASE mode does not compile with CLANG. Compilation breaks with error `.././src/base/atomicops_internals_mips64_gcc.h:234:25: error: instruction requires a CPU feature not currently enabled
                       "lld %0, %2\n"  // temp = *ptr
                        ^
`

This commit updated the version of CLANG in dependencies, so the problem is probably that CLANG misses a flag on the command line (or maybe has an extra unneeded flag).

Have a look and see what the problem is.

**Comment by ibogosavljevic** [comment]
Issue reported to CLANG team. They have confirmed the bug, here is the issue:
https://dmz-portal.mips.com/bugz/show_bug.cgi?id=2240

**Comment by ibogosavljevic** [comment]
Here are the updates of the compiler team:
They say that the error message here `instruction requires a CPU feature not currently enabled` is not correct, and they will fix it. But, alas, this doesn't fix our issue. The format of LLD instruction has changed  between r2 and r6. On r6, the offset field is just 9 bits long, whereas it is 16 bits long on r2. What we have here is that the instruction cannot be assembler for r6 because the offset parameter is too long (e.g. lld 
a0, 550(a1) works on r2 but not on r6).
Most of the times we can fetch the argument using 9 bits only, but depending on circumstances, this doesn't work all the time. When it doesn't work, the compiler complains. The guys from the compiler team were unable to tell me exactly if it is compiler's job to make address fit, or it is our job, but they recommend we load the address with LD first, and then access it using LLD and offset 0.

**Comment by petar-jovanovic** [comment]
I will have to agree that a recommended approach in this particular case is to change the inline assembler in the way to use a value from a known register that gets loaded before the inline block, i.e. we let compiler set the pointer to the memory in a correct way.

**Comment by paul99** [comment]
In https://dmz-portal.mips.com/bugz/show_bug.cgi?id=1869 this was fixed in gcc/binutils, to correctly use the inline assembly that we have now. Should we not have clang/integrated-as work the same?

**Comment by petar-jovanovic** [comment]
Indeed. Let me elaborate this with a few more words. First of all, I do _not_ have a strong opinion on this, I had it before when I hoped we can easily fix the compiler.
However, what we did with gcc/binutils (thanks for bringing the issue back, I completely forgot about it and how we approached it) is to introduce a new assembly macro. There are good arguments that say we should not be adding more assembly macros. Macros that expand into multiple instructions are often source of issues, etc. LLVM's integrated assemblers is, on the other hand, envisioned as a solid yet modest feature-wise assembler that should support common cases, but not every single boundary case available in gas (hence one more reason to question if a support for a new macro should be added). The problem we see in this issue is not R6 specific, we should be able to reproduce it with a different example on R2. 'lld' has never been a macro, and now it is (hmm, how will engineers find about it?). I would say the inline assembler should be written in a different way, so the assembler part stays as close to what we expect as possible. For that reason, I'd advocate a change, but again - I do not have strong opinion on this anymore.
Having said this all, I actually encouraged one of the engineers today to add this feature into LLVM IAS, so in the end, the arguments for a change are even weaker.

**Comment by paul99** [comment]
@petar-jovanovic -  from above: 'I actually encouraged one of the engineers today to add this feature into LLVM IAS'. Do you know if this feature was or will be implemented?

**Comment by ibogosavljevic** [comment]
Here is the patch which workarounds the problem with non-compilation on CLANG using R6. Tested it and it does not introduce new failures (compared with debug version since optdebug doesn't compile without the patch)

```
diff --git a/src/base/atomicops_internals_mips64_gcc.h b/src/base/atomicops_internals_mips64_gcc.h
index 85b4e46..19ff439 100644
--- a/src/base/atomicops_internals_mips64_gcc.h
+++ b/src/base/atomicops_internals_mips64_gcc.h
@@ -90,19 +90,21 @@ inline Atomic32 NoBarrier_AtomicExchange(volatile Atomic32* ptr,
 inline Atomic32 NoBarrier_AtomicIncrement(volatile Atomic32* ptr,
                                           Atomic32 increment) {
   Atomic32 temp, temp2;
+  volatile void * x = (void *)(ptr);

   __asm__ __volatile__(".set push\n"
                        ".set noreorder\n"
+                       "move $t0, %5\n"
                        "1:\n"
-                       "ll %0, %2\n"  // temp = *ptr
+                       "ll %0, 0($t0)\n"  // temp = *ptr
                        "addu %1, %0, %3\n"  // temp2 = temp + increment
-                       "sc %1, %2\n"  // *ptr = temp2 (with atomic check)
+                       "sc %1, 0($t0)\n"  // *ptr = temp2 (with atomic check)
                        "beqz %1, 1b\n"  // start again on atomic error
                        "addu %1, %0, %3\n"  // temp2 = temp + increment
                        ".set pop\n"
                        : "=&r" (temp), "=&r" (temp2), "=m" (*ptr)
-                       : "Ir" (increment), "m" (*ptr)
-                       : "memory");
+                       : "Ir" (increment), "m" (*ptr), "r" (x)
+                       : "t0", "memory");
   // temp2 now holds the final value.
   return temp2;
 }
@@ -227,19 +229,21 @@ inline Atomic64 NoBarrier_AtomicExchange(volatile Atomic64* ptr,
 inline Atomic64 NoBarrier_AtomicIncrement(volatile Atomic64* ptr,
                                           Atomic64 increment) {
   Atomic64 temp, temp2;
+  void * x = (void *) ptr;

   __asm__ __volatile__(".set push\n"
                        ".set noreorder\n"
+                       "move $t0, %5\n"
                        "1:\n"
-                       "lld %0, %2\n"  // temp = *ptr
+                       "lld %0, 0($t0)\n"  // temp = *ptr
                        "daddu %1, %0, %3\n"  // temp2 = temp + increment
-                       "scd %1, %2\n"  // *ptr = temp2 (with atomic check)
+                       "scd %1, 0($t0)\n"  // *ptr = temp2 (with atomic check)
                        "beqz %1, 1b\n"  // start again on atomic error
                        "daddu %1, %0, %3\n"  // temp2 = temp + increment
                        ".set pop\n"
                        : "=&r" (temp), "=&r" (temp2), "=m" (*ptr)
-                       : "Ir" (increment), "m" (*ptr)
-                       : "memory");
+                       : "Ir" (increment), "m" (*ptr), "r" (x)
+                       : "t0", "memory");
   // temp2 now holds the final value.
   return temp2;
 }
```

**Comment by MIPSbkirby** [comment]
The above patch is now being applied to the "Target - MIPS64 - edge - clang,mipsr6 - builder" compiler builder.  I've run that builder and Clang now successfully compiles V8.

**Comment by gordana-cmiljanovic-imgtec** [comment]
I came across our old lld/scd change maybe it can be of use here: https://mipsia.review.mips.com/#/c/2723/

This one was uploaded to chromium_org in android, but it should apply to v8 as well.
